src/big/big_decimal.cr

```diff
@@ -0,0 +1,241 @@
+    if @scale > other.scale
+      scaled = other.scale_to(self)
+      BigDecimal.new(@value + scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value + other.value, other.scale)
+    else
+      BigDecimal.new(@value + other.value, @scale)
+    end
+  end
+
+  def -(other : BigDecimal) : BigDecimal
+    if @scale > other.scale
+      scaled = other.scale_to(self)
+      BigDecimal.new(@value - scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value - other.value, other.scale)
+    else
+      BigDecimal.new(@value - other.value, @scale)
+    end
+  end
+
+  def *(other : BigDecimal) : BigDecimal
+    BigDecimal.new(@value * other.value, @scale + other.scale)
+  end
+
+  def /(other : BigDecimal) : BigDecimal
+    check_division_by_zero other
+
+    scale = @scale - other.scale
+    n, d = @value, other.@value
+
+    quotient, remainder = n.tdiv(d), n.remainder(d)
+    return BigDecimal.new(quotient, scale) if remainder == ZERO
+
+    remainder = remainder * TEN
+
+    i = 0
+    while remainder != ZERO && i < @@max_div_iterations
+      q, r = remainder.tdiv(d), remainder.remainder(d)
+      quotient = quotient * TEN + q
+      remainder = r * TEN
+      i += 1
+    end
+
+    if (@value < 0 && other.value > 0) || (other.value < 0 && @value > 0)
+      BigDecimal.new(-quotient, scale + i)
+    else
+      BigDecimal.new(quotient, scale + i)
+    end
+  end
+
+  def <=>(other : BigDecimal) : Int32
+    if @scale > other.scale
+      @value <=> other.scale_to(self).value
+    elsif @scale < other.scale
+      scale_to(other).value <=> other.value
+    else
+      @value <=> other.value
+    end
+  end
+
+  def <=>(other : Int)
+    @value <=> other
+  end
+
+  def ==(other : BigDecimal) : Bool
+    if @scale > other.scale
+      scaled = other.value * power_ten_to(@scale - other.scale)
+      @value == scaled
+    elsif @scale < other.scale
+      scaled = @value * power_ten_to(other.scale - @scale)
+      scaled == other.value
+    else
+      @value == other.value
+    end
+  end
+
+  # Scales a `BigDecimal` to another `BigDecimal`, so they can be
+  # computed easier.
+  def scale_to(new_scale : BigDecimal) : BigDecimal
+    new_scale = new_scale.scale
+
+    if @value == 0
+      BigDecimal.new(0.to_big_i, new_scale)
+    elsif @scale > new_scale
+      scale_diff = @scale - new_scale.to_big_i
+      BigDecimal.new(@value / power_ten_to(scale_diff), new_scale)
+    elsif @scale < new_scale
+      scale_diff = new_scale - @scale.to_big_i
+      BigDecimal.new(@value * power_ten_to(scale_diff), new_scale)
+    else
+      self
+    end
+  end
+
+  def to_s(io : IO)
+    s = @value.to_s
+
+    if @scale == 0
+      io << s
+      return
+    end
+
+    if @scale == s.size
+      io << "0." << s
+    elsif @scale > s.size
+      io << "0."
+      (@scale - s.size).times do
+        io << "0"
```

Comment:
You could use a `Char` here.

Reply:
No, string is better here because need no conversion from Char to String.

Reply:
@Sija I'm wrong! Unbelievable! 

```
➜  crystal git:(bigfloat-frexp) ✗ bin/crystal 1.cr --no-debug --release
Using compiled compiler at `.build/crystal'
            char 370.13  (   2.7ms) (± 0.91%)       fastest
          string 150.32  (  6.65ms) (± 1.97%)  2.46× slower
  byte from char  12.61  ( 79.32ms) (± 2.65%) 29.36× slower
byte from string 133.63  (  7.48ms) (± 1.40%)  2.77× slower
            byte 132.72  (  7.53ms) (± 1.82%)  2.79× slower
```

```crystal
require "benchmark"

Benchmark.ips do |x|
  x.report("char") { io = IO::Memory.new; 1_000_000.times { io << '.' } }
  x.report("string") { io = IO::Memory.new; 1_000_000.times { io << "." } }
  x.report("byte from char") { io = IO::Memory.new; 1_000_000.times { io << '.'.bytes } }
  x.report("byte from string") { io = IO::Memory.new; 1_000_000.times { io << ".".byte_at(0) } }
  x.report("byte") { b = ".".byte_at(0); io = IO::Memory.new; 1_000_000.times { io << b } }
end
```

Reply:
@vegai Could you please include above tweaks?

Reply:
Huh, interesting :) Seems almost like a performance bug somewhere.

But will change these to chars then.